tests/unit/adapters/osf-adapter-test.js

```diff
@@ -1,12 +1,401 @@
+    this.stub(node, 'resolveRelationship', () => {
+        return {
+            addCanonicalRecord: addCanonicalStub
+        };
+    });
+
+    Ember.run(() => {
+        node.save().then(() => {
+            saveStubs.forEach(s => assert.ok(s.called));
+            assert.ok(addCanonicalStub.calledTwice);
+            // Can't use calledWith because sinon's deepEqual creates
+            // infinite recursive calls when comparing the Ember DS.Models
+            assert.equal(addCanonicalStub.args[0][0], contributors[0]);
+            assert.equal(addCanonicalStub.args[1][0], contributors[1]);
+        }, () => {
+            // Fail
+            assert.ok(false);
+        });
+    });
+});
+
+test('#_addRelated defers to _doRelatedRequest and adds records to the parent\'s canonical state', function(assert) {
+    let node = FactoryGuy.make('node');
+    let institution = FactoryGuy.make('institution');
+    node.get('affiliatedInstitutions').pushObject(institution);
+
+    var doRelatedStub = this.stub(OsfAdapter.prototype, '_doRelatedRequest', () => {
+        return new Ember.RSVP.Promise(resolve => resolve());
+    });
+    var relation = node.resolveRelationship('affiliatedInstitutions');
+    relation.hasLoaded = true;
+    var addCanonicalStub = this.stub(relation, 'addCanonicalRecord');
+
+    Ember.run(() => {
+        node.save().then(() => {
+            assert.ok(doRelatedStub.called);
+            assert.ok(addCanonicalStub.calledOnce);
+            assert.ok(addCanonicalStub.calledWith(institution));
+        }, () => {
+            // Fail
+            assert.ok(false);
+        });
+    });
+});
+
+test('#_updateRelated defers to _doRelatedRequest, pushes the update response into the store, and updates the parent\'s canonicalState', function(assert) {
+    this.inject.service('store');
+    let store = this.store;
+
+    let node = FactoryGuy.make('node', 'hasContributors');
+    var contribs = node.get('contributors');
+    var contrib = contribs.objectAt(1);
+
+    contrib.set('bibliographic', !contrib.get('bibliographic'));
+
+    var doRelatedStub = this.stub(OsfAdapter.prototype, '_doRelatedRequest', () => {
+        return new Ember.RSVP.Promise(resolve => resolve({
+            data: [
+                // A slight hack-- ingore the value returned from _doRelatedRequest
+                true
+            ]
+        }));
+    });
+    var addCanonicalStub = this.stub();
+    this.stub(node, 'resolveRelationship', () => {
+        return {
+            addCanonicalRecord: addCanonicalStub
+        };
+    });
+    var pushStub = this.stub(store, 'push', () => contrib);
+    var normalizeStub = this.stub(store, 'normalize');
+
+    Ember.run(() => {
+        node.save().then(() => {
+            assert.ok(doRelatedStub.calledOnce);
+            assert.ok(addCanonicalStub.calledOnce);
+            assert.ok(pushStub.calledOnce);
+            assert.ok(normalizeStub.calledOnce);
+        }, () => {
```

Comment:
Instead of passing two arguments to `.then`, the second failure handler can be specified as `.then().catch()`. My impression is that using separate handlers is preferred over using two arguments to `.then`, partly because it helps ensure that promises are chained correctly- eg to ensure that `.catch` handles errors in the first request *or* any that happen during the success callback.

(not 100% sure about this, as some articles are a bit old)